src/sinks/util/sink.rs

```diff
@@ -0,0 +1,1193 @@
+//! This module contains all our internal sink utilities
+//!
+//! All vector "sinks" are built around the `Sink` type which
+//! we use to "push" events into. Within the different types of
+//! vector "sinks" we need to support three main use cases:
+//!
+//! - Streaming sinks
+//! - Single partition batching
+//! - Multiple partition batching
+//!
+//! For each of these types this module provides one external type
+//! that can be used within sinks. The simplest type being the `StreamSink`
+//! type should be used when you do not want to batch events but you want
+//! to _stream_ them to the downstream service. `BatchSink` and `PartitonBatchSink`
+//! are similar in the sense that they both take some `tower::Service`, `Batch` and
+//! `Acker` and will provide full batching, request dipstaching and acking based on
+//! the settings passed.
+//!
+//! For more advanced use cases like http based sinks, one should use the
+//! `BatchedHttpSink` type, which is a wrapper for `BatchSink` and `HttpSink`.
+//!
+//! # Driving to completetion
+//!
+//! Each sink utility provided here strictly follows the patterns described in
+//! the `futures01::Sink` docs. Each sink utility must be polled from a valid
+//! tokio context wether that may be an actual runtime or using any of the
+//! `tokio01-test` utilities.
+//!
+//! For service based sinks like `BatchSink` and `PartitionBatchSink` they also
+//! must be polled within a valid tokio executor context or passed a valid executor.
+//! This is due to the fact that they will spawn service requests to allow them to be
+//! driven independently from the sink. A oneshot channel is used to tie them back into
+//! the sink to allow it to notify the consumer that the request has succeeded.
+
+use super::batch::{Batch, BatchSettings};
+use super::buffer::partition::Partition;
+use crate::buffers::Acker;
+use futures01::{
+    future::Either,
+    stream::FuturesUnordered,
+    sync::oneshot::{self, Receiver},
+    try_ready, Async, AsyncSink, Future, Poll, Sink, StartSend, Stream,
+};
+use std::{
+    collections::{HashMap, VecDeque},
+    fmt,
+    hash::Hash,
+    marker::PhantomData,
+    time::Instant,
+};
+use tokio01::{
+    executor::{DefaultExecutor, Executor},
+    timer::Delay,
+};
+use tower::Service;
+use tracing_futures::Instrument;
+
+// === StreamSink ===
+
+const STREAM_SINK_MAX: usize = 10_000;
+
+/// Simple stream based sink adapter.
+///
+/// This will wrap any inner sink acking all events
+/// as soon as poll_complete returns ready. `start_send`
+/// will also attempt to fully flush if the amount of
+/// in flight acks is larger than `STREAM_SINK_MAX`.
+#[derive(Debug)]
+pub struct StreamSink<T> {
+    inner: T,
+    acker: Acker,
+    pending: usize,
+}
```

Comment:
Just to be clear, this is a basically a renamed `StreamAck`?

Reply:
Yes.